[mlir][spirv] Add canon patterns for IAddCarry/[S|U]MulExtended

inbelic opened the pull request:
Add missing constant propogation folder for IAddCarry and [S|U]MulExtended. Due to currently missing constant value for spirv.struct the folding is done using canonicalization patterns.

Implement additional folding when rhs is 0 for all ops and when rhs is 1 for UMulExt.

This helps for readability of lowered code into SPIR-V.

Part of work for #70704

inbelic committed:
[mlir][spirv] Add folding for IAddCarry/[S|U]MulExtended

Add missing constant propogation folder for IAddCarry and
[S|U]MulExtended. Due to currently missing constant value for
spirv.struct the folding is done using canonicalization patterns.

Implement additional folding when rhs is 0 for all ops and when rhs
is 1 for UMulExt.

This helps for readability of lowered code into SPIRV.

Part of work for #70704

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -115,6 +115,196 @@ void spirv::AccessChainOp::getCanonicalizationPatterns(
+  LogicalResult matchAndRewrite(spirv::IAddCarryOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    auto operands = op.getOperands();
```
We can do:
```c++
Value lhs = op.getOperand1();
Value rhs = op.getOperand2();
```

If this is not a very intuitive naming, I can add extra getters `op.getLhs()` and `op.getRhs()` -- WDYT?

inbelic replied:
Agree this looks better. I think it is okay to be done locally rather than having the getters.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -115,6 +115,196 @@ void spirv::AccessChainOp::getCanonicalizationPatterns(
+
+    auto adds = constFoldBinaryOp<IntegerAttr>(
+        {lhs, rhs}, [](const APInt &a, const APInt &b) { return a + b; });
+    if (!adds)
+      return failure();
+
+    auto carrys = constFoldBinaryOp<IntegerAttr>(
+        ArrayRef{adds, lhs}, [](const APInt &a, const APInt &b) {
+          APInt zero = APInt::getZero(a.getBitWidth());
+          return a.ult(b) ? (zero + 1) : zero;
+        });
+
+    if (!carrys)
+      return failure();
```
I think that instead of using `CompositeConstruct` we could also create two insertions into the struct -- one for the sum, a separate one for the carry. This way we don't have to be able to fold both to simplify the result.

For example when we see something like this:
```
%x = spirv.UConvert %a : i8 to i32
%y = spirv.UConvert %b : i8 to i32
%res = spirv.IAddCarry %a, %b : i32
```
Here we can statically tell that the overflow bit is `0`, but can't necessarily decide that the sum component is.

This may be a bit obscure, but maybe we could add it as a TODO comment.

inbelic replied:
I agree that if it can allow for more greedy folding then we should use it. We also have the `-spirv-rewrite-inserts` pass that can convert from a `CompositeInsert` chain to the `CompositeConstruct` in the cases when desired.

The most recent commit has a potential implementation of switching to this.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -115,6 +115,196 @@ void spirv::AccessChainOp::getCanonicalizationPatterns(
+        rewriter.create<spirv::ConstantOp>(loc, constituentType, lowBits);
+    constituents.push_back(lowBitsVal);
+
+    Value highBitsVal =
+        rewriter.create<spirv::ConstantOp>(loc, constituentType, highBits);
+    constituents.push_back(highBitsVal);
```
Also here: we could potentially fold the high bits if the operands are known to be small enough.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -115,6 +115,196 @@ void spirv::AccessChainOp::getCanonicalizationPatterns(
+    constituents.push_back(carrysVal);
+
+    rewriter.replaceOpWithNewOp<spirv::CompositeConstructOp>(op, op.getType(),
+                                                             constituents);
```
I think we can replace the `constituents` vector with a simple stack array:
```c++
Value constituents[2] = {addVals, carrysVal};
```

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -115,6 +115,196 @@ void spirv::AccessChainOp::getCanonicalizationPatterns(
+    constituents.push_back(highBitsVal);
+
+    rewriter.replaceOpWithNewOp<spirv::CompositeConstructOp>(op, op.getType(),
+                                                             constituents);
```
Same here

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -115,6 +115,196 @@ void spirv::AccessChainOp::getCanonicalizationPatterns(
+    // umulextended (x, 1) = <x, 0>
+    if (matchPattern(operands[1], m_One())) {
+      Value zero = spirv::ConstantOp::getZero(constituentType, loc, rewriter);
+      constituents.push_back(operands[0]);
+      constituents.push_back(zero);
+      rewriter.replaceOpWithNewOp<spirv::CompositeConstructOp>(op, op.getType(),
```
Also here

kuhar commented on mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir:
```diff
@@ -336,6 +336,52 @@ func.func @iadd_poison(%arg0: i32) -> i32 {
+  %cn5 = spirv.Constant -5 : i32
+  %cn8 = spirv.Constant -8 : i32
+
+  // CHECK-DAG: spirv.Constant 0
+  // CHECK-DAG: spirv.Constant -3
+  // CHECK-DAG: spirv.CompositeConstruct
+  // CHECK-DAG: spirv.Constant 1
+  // CHECK-DAG: spirv.Constant -13
+  // CHECK-DAG: spirv.CompositeConstruct
```
Same comment as on the other PR: we should make sure that these are used like we'd expect

inbelic committed:
review comments

- improve readability with lhs/rhs instead of operands[0]/[1]
- use stack array instead of llvm::SmallVector
- increase strictness of tests to ensure proper CompositeConstruct and
  return order

inbelic committed:
switch to using CompositeInsert from Construct

- commit to demonstrate how we could potentially use CompositeInsert
  instead of CompositeConstruct

kuhar reviewed (approved):
LGTM, thanks!

kuhar commented:
BTW @inbelic, with this many PRs in you should be able to request write access if you haven't done so already. The instructions are here: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access
